### PR TITLE
Script-Extension: RuleSupport

### PR DESCRIPTION
**Pull request by smerschjohann**
This Script Extension allows to define Rules, ModuleHandlers and ModuleTypes by script. 

Still open:

OSGI-Bundle order handling: The RuleRegistry has a bad behavior when the ManagedProvider is not available, it does not allow to add rules in that case: "ManagedProvider is not available". This happens directly after start. 

Maybe Karaf is a solution?

# Example
Here is an example of how you can define rules, in addition to this there exists also some other ways to define rules and modules. This needs to be documented in the future.

```
print ScriptExtension.presets
 
ScriptExtension.importPreset("RuleSupport")
ScriptExtension.importPreset("RuleSimple")
  

def itemStateChangeTrigger(triggername, itemName):
    return Trigger(triggername, "ItemStateChangeTrigger", {
        "itemName": itemName
    })
 
import org.eclipse.smarthome.automation.handler.ActionHandler as ActionHandler
	
class MyAction(ActionHandler):
	def __init__(self, module):
		self.module = module
		
	def execute(self, context):
		print "received context", context
		return {"result": "awesome"}

		
import org.eclipse.smarthome.automation.type.ActionType as ActionType
import org.eclipse.smarthome.config.core.ConfigDescriptionParameter as ConfigDescriptionParameter

HandlerRegistry.addActionType(ActionType("PyTestAction", [ConfigDescriptionParameter("string",ConfigDescriptionParameter.Type.TEXT)], []))	
HandlerRegistry.addActionHandler("PyTestAction", lambda module: MyAction(module)) 
 
# SimpleRule can combine any TriggerType and ConditionType registered to the RuleEngine
class MySimpleRule(SimpleRule):
    def __init__(self):
        self.actions.add(Action("test", "PyTestAction", {"string": "hopefully working"}, None))
        self.triggers = [itemStateChangeTrigger("trigger1", "Light_C_Corridor_Ceiling")]

    def execute(self, module, inputs):
        print "module:", module
        print "configuration:", module.configuration
        print "inputs:", inputs

        print dir(inputs)
        for k, v in inputs.iteritems():
            print "key:", k, "value:", v
			


HandlerRegistry.addRule(MySimpleRule())        
```

### CODE REVIEW

**Comment by danchom** [comment]
We have a long discussion about this behavior. The properties of rule, its configuration, triggers, configuration, triggers, condition and actions has to be store persistently, because the changes in the rule must be available between restarts of framework. The persistence is done by managed provider. 
If a rule is created and configured without ManagedProvider and the framework is restarted, the changes will be lost (these changes can be done by someone other then the creator of the rule). The other pure case is if you create a new rule and the managed provider has a stored rule with the same uid, but not loaded yet. For that reason the RuleEngine first loads stored rules form managed provider before permitting to add new once.
Is it possible this script-extention to listen to RuleRegistry and to be registered after that?

**Comment by kaikreuzer** [comment]
How did we solve the startup order in the RuleImporter? This also tries adding rules on startup and thus requires the ManagedProvider to be present already.

File: bundles/automation/org.eclipse.smarthome.automation.module.script.extensions.rulesupport/ESH-INF/automation/moduletypes/PrivateScriptedTypes.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
+{
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.module.script.extensions.rulesupport/ESH-INF/automation/moduletypes/PrivateScriptedTypes.json]: Why do you need these types and cannot simply use https://github.com/eclipse/smarthome/blob/master/bundles/automation/org.eclipse.smarthome.automation.module.script/ESH-INF/automation/moduletypes/ScriptTypes.json?

Comment (smerschjohann) [inline bundles/automation/org.eclipse.smarthome.automation.module.script.extensions.rulesupport/ESH-INF/automation/moduletypes/PrivateScriptedTypes.json]: The implementation of these private type functions are bound be the ScriptEngine, they cannot simply be serialized and later on evaluated by ScriptAction, ScriptCondition etc.. Here I save the function handler in combination with a "privId".

**Comment by danchom** [comment]
The RuleImporter is waiting for Manege Rule Provider registration and after that tries to add new rules. 
We have discussed  that the rule has to continue to work even its module type is removed. Unfortunately it can't work without handler of the module type. For that reason we decided not to separate module type from its handler. But it is not good if someone uses a third party module type into a rule and this type is removed. When the type is disappeared the rule goes into not initialized state but this can brake application logic of the rule creator.

**Comment by smerschjohann** [comment]
> But it is not good if someone uses a third party module type into a rule and this type is removed. When the type is disappeared the rule goes into not initialized state but this can brake application logic of the rule creator.

What does this mean? Yes, it should result in "uninitialize" as the rule as a whole can not function if a single part is missing without possible change in behavior. 

I build this extension in a way, that, by default, rules that are registered by a script are removed when the script is unloaded. Because of this, it will end in a known state before and after (final) script execution. But if a user wants, he can add a rule permanently (by using addPermanent) or add them by a rule creator in combination with named ModuleHandlers.

But why would removing handlers break application logic of a rule creator? I don't get really get that, can you explain it a bit?


> The RuleImporter is waiting for Manege Rule Provider registration and after that tries to add new rules.

There are too many indirections for my taste:

- The RuleRegistry has the dependency on ManagedRuleProvider, without it, it is not functioning. Wouldn't it be better in that case if the RuleRegistry is only registered to OSGI when the ManagedRuleProvider is available?  At least I don't see why a bundle that has a dependency on the RuleRegistry, should have the need to check for dependencies the RuleRegistry itself has.
- The ScriptExtension here would only depend on the RuleRegistry, but because of the current behavior, also on the ManagedRuleProvider.
- The ScriptLoader again depends on ScriptExtensions which should normally be available before script execution, otherwise this could lead to Exceptions in the scripts. So in general it would be a good idea to have some mechanism to at least be able to sort the start order in some way. There is another workaround by delaying the script execution on boot and try executing scripts, if they hit an exception, retry after x seconds. Not very nice.. 

As the script loader cannot know which ScriptExtensions will still be loaded, it cannot wait for them accordingly. As you are currently introducing Karaf, it might be easier to use feature files for this? I don't know Karaf until @kaikreuzer mentioned it, but I've read that it could solve these kind of issues. 

It would mean here:
- the RuleRegistry is loaded after the ManagedRuleProvider or at least register itself after it received the ManagedRuleProvider.
- all bundles that depend on the RuleRegistry, will be loaded after the RuleRegistry is fully functional or at least the RuleRegistry is only set by OSGI if the RuleRegistry is functional..
- the ScriptLoader is ordered after the start of all ScriptExtensions that should be loaded on initialization

**Comment by kaikreuzer** [comment]
> As you are currently introducing Karaf

No, we are not. Karaf is only used in openHAB 2 and ESH comes with some helpful meta-data, but ESH solutions are not at all tied to using Karaf - hence the code must not have any dependency on it.

**Comment by smerschjohann** [comment]
ok, but I don't see a real solution, how a bundle itself can resolve this issue.

**Comment by kaikreuzer** [comment]
I didn't dive into the details of the dependencies you are listing here, but in general this is how OSGi works - every service should clearly define which service he depends on and it will only be started if all of them are resolvable. So if you have a service, that requires rules to be persisted by the framework, it should depend on the ManagedRuleProvider (even if not directly calling it directly in the code).

**Comment by smerschjohann** [comment]
Is it in addition possible for a bundle to iterate through OSGI bundles that are still to load and check if they provide a special service (in this case a ScriptExtension)?

**Comment by kaikreuzer** [comment]
This conceptionally could never work as you might install an additional bundle an hour later (and is thus "still to be loaded"), so this is totally unknown during startup time.

**Comment by smerschjohann** [comment]
Therefore I said still to load or in other words "in the queue" ;) Yes you can install bundles at any time, but there is a difference between "I am in a known bootup" and "I might install additional bundles". If again looking at ScriptExtensions: On the one hand you might have ScriptExtensions already "installed" locally and depend on them in your scripts. These scripts would not work without them, so it would be best to wait for these bundles.
On the other hand you have completely new bundles, which in this case no script relies on so far, so you don't have to wait for them and if installed, they can be instantly used in the scripts. So by adding or modifying the scripts, they can now use the extensions - everything is fine.

If one can find out that OSGI is finished with its initial startup would be enough already, otherwise we are back to Thread.sleep()...

**Comment by danchom** [comment]
>But why would removing handlers break application logic of a rule creator? I don't get really get that, >can you explain it a bit?
For example if you have a rule which contains two conditions and a handler for some of them is missing. Rule execution require both condition to be satisfied and if one missing then this is not same rule. 
If the rule executes more then one action and some of them can be executed because they does not have handlers the entire system may go into not determinate state. 
The same is with triggers, the rule relay on  two trigger, but it is triggered only from one of them because the other does not have a handler.
In the real system we have two different roles:
Developer - which creates module types and handlers for them, and SysAdmin - the person which create rules and relay on module type and handlers available in the system. For that reason IMHO the module types are something static. When a developer is a sysAdmin this is a corner case. Only the sysAdmin has ability to change the rules. Usually the sysAdmin is the owner of the home (which is not developer). 

>The RuleRegistry has the dependency on ManagedRuleProvider, without it, it is not functioning. 
RuleRegistry does not have dependency to the ManagedRuleProvider because there should be other rule provider which can produce rules, but these rules are unmodifiable and appears and disappeared with their providers.  The RuleRegistry has to return these rules. 
You can define your script extention as rule provider instead of part of the script handler. But if you want to create a new rules through the RuleRegistry you have to wait util ManagedRuleProvider to be registered.

**Comment by kgoderis** [comment]
@smerschjohan Simon, I suppose this is what you referred to in https://github.com/eclipse/smarthome/pull/803 ? I see little overlap with the code I am brewing, I rather do all my modifications on the "inside". I only see small overlaps with what javax.script is already providing (ScriptEngineManager and alike). When it is a bit cleaner I will push the my repo and share.

**Comment by smerschjohann** [comment]
This PR is superseded by https://github.com/eclipse/smarthome/issues/1783
